### PR TITLE
Use only subm cluster to recover broker-info.subm

### PR DESCRIPTION
**Pull request by Jaanki**
Epic: https://github.com/submariner-io/enhancements/issues/143

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: cmd/subctl/recover_broker_info.go [hunk 1]
```diff
@@ -63,114 +55,57 @@ func init() {
...
+	brokerNamespace := submCluster.Submariner.Spec.BrokerK8sRemoteNamespace
+	brokerRestConfig := submCluster.RestConfig
 
-	clusters, err := brokerCluster.GetClusters(brokerNamespace)
+	status.Start("Checking if the Broker is installed on the Submariner cluster %q in namespace %q", submCluster.Name, brokerNamespace)
+
+	brokerObj, err := getBroker(brokerRestConfig, brokerNamespace)
 	if err != nil {
-		return status.Error(err, "error listing joined clusters")
+		status.Warning("Error getting Broker: %s", err)
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]: An error should be logged as such and abort the operation.

```suggestion
		return status.Error(err, "")
```

File: cmd/subctl/recover_broker_info.go [hunk 2]
```diff
@@ -63,114 +55,57 @@ func init() {
...
 	}
 
-	ok, err := tryToRecoverFromBroker(brokerCluster, brokerObj, brokerNamespace, clusters, status)
-	if ok || err != nil {
-		return err
+	if brokerObj == nil {
+		status.Start("Trying to connect to the Broker installed on a different cluster")
+
+		brokerRestConfig, brokerNamespace, err = restconfig.ForBroker(submCluster.Submariner, submCluster.ServiceDiscovery)
+		if err != nil || brokerRestConfig == nil {
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]: `brokerRestConfig` won't be nil if there's no error.

File: cmd/subctl/recover_broker_info.go [hunk 3]
```diff
@@ -63,114 +55,57 @@ func init() {
...
+			return status.Error(err, "Error getting the broker's rest config")
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]:
```suggestion
			return status.Error(err, "Error getting the Broker's REST config")
```

File: cmd/subctl/recover_broker_info.go [hunk 4]
```diff
@@ -63,114 +55,57 @@ func init() {
...
+		}
+
+		brokerObj, err = getBroker(brokerRestConfig, brokerNamespace)
+		if err != nil {
+			return status.Error(err, "Error getting Broker")
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]: The error already contains enough info.

```suggestion
			return status.Error(err, "")
```

File: cmd/subctl/recover_broker_info.go [hunk 5]
```diff
@@ -63,114 +55,57 @@ func init() {
...
+		}, brokerObj)
 
-func getBroker(clusterInfo *cluster.Info, namespace, notFoundMsg string, status reporter.Interface) (*v1alpha1.Broker, error) {
-	status.Start("Checking if the Broker is installed on cluster %q in namespace %q", clusterInfo.Name, namespace)
-
-	brokerObj, err := clusterInfo.GetBroker(namespace)
 	if apierrors.IsNotFound(err) {
-		return nil, status.Error(fmt.Errorf("the Broker is not installed on the specified cluster in namespace %s. %s", notFoundMsg,
-			namespace), "")
+		return nil, errors.Wrapf(err, "the Broker is not installed on the specified cluster in namespace %s", namespace)
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]: This is not an error condition. We should return nil error. The `nilnil` linter may complain about this in which case return a `found` bool.

File: cmd/subctl/recover_broker_info.go [hunk 6]
```diff
@@ -63,114 +55,57 @@ func init() {
...
+
+		brokerRestConfig, brokerNamespace, err = restconfig.ForBroker(submCluster.Submariner, submCluster.ServiceDiscovery)
+		if err != nil || brokerRestConfig == nil {
+			return status.Error(err, "Error getting the broker's rest config")
+		}
+
+		brokerObj, err = getBroker(brokerRestConfig, brokerNamespace)
+		if err != nil {
+			return status.Error(err, "Error getting Broker")
+		}
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go:82]: We should also check if the `brokerObj` isn't found. ACM doesn't create a `Broker` resource and neither does helm I believe (currently anyway).

```suggestion
		}
		
		if brokerObj == nil {
		    return status.Error(fmt.Errorf("No Broker resource found in namespace %q", brokerNamespace), "")
		}
```

File: cmd/subctl/recover_broker_info.go [hunk 7]
```diff
@@ -63,114 +55,57 @@ func init() {
...
 	}
 
-	status.Warning("Submariner is not installed on the same cluster as Broker")
-	status.Start("Checking if Submariner is installed on a different cluster")
+	status.Success("Found Broker installed on a different cluster in namespace %s", brokerNamespace)
```
Comment (tpantelis) [inline cmd/subctl/recover_broker_info.go]: This should be inside the `if` block above. Also we need to end the status here.

```suggestion
	status.End()
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: cmd/subctl/recover_broker_info.go (hunk 1 above)
Comment (Jaanki) [reply cmd/subctl/recover_broker_info.go]: We don't want to return here. This addresses the case where Broker is not installed on the current context and hence we move ahead to getting remote Brokerconfig and get BrokerObj from that remote cluster.

File: cmd/subctl/recover_broker_info.go (hunk 6 above)
Comment (Jaanki) [reply cmd/subctl/recover_broker_info.go:82]: We don't intent to use `recover-broker-info` command with ACM anyways. However, I will make the suggested changes.

File: cmd/subctl/recover_broker_info.go (hunk 3 above)
Comment (Jaanki) [reply cmd/subctl/recover_broker_info.go]: All other usages have lower cases. Hence I used lower case here. Anyways, I will change it.

File: cmd/subctl/recover_broker_info.go (hunk 1 above)
Comment (Jaanki) [reply cmd/subctl/recover_broker_info.go]: Changed the return err in L103 to be able to return here.

File: cmd/subctl/recover_broker_info.go (hunk 6 above)
Comment (tpantelis) [reply cmd/subctl/recover_broker_info.go:82]: Right but we can't control what a user might do 😄

File: cmd/subctl/recover_broker_info.go (hunk 1 above)
Comment (tpantelis) [reply cmd/subctl/recover_broker_info.go]: yeah this goes along with my comment in `getBroker` - I should've been clear about that. Basically we need to distinguish between "not found" and an unrecoverable error and, for the latter, abort. You could also just have `getBroker` propagate the error as is and check `apierrors.IsNotFound` here.

File: cmd/subctl/recover_broker_info.go (hunk 3 above)
Comment (tpantelis) [reply cmd/subctl/recover_broker_info.go]: yeah it's fine either way. We should probably be consistent with "Broker" vs "broker" even tho this is an error message and won't normally be seen.

File: cmd/subctl/recover_broker_info.go (hunk 5 above)
Comment (tpantelis) [reply cmd/subctl/recover_broker_info.go]: An alternative is to propagate the error as is and check `apierrors.IsNotFound` in the caller. I think that's more idiomatic.

**Review by tpantelis — APPROVED** [review]

**Review by maayanf24 — APPROVED** [review]

**Pull request merged**